### PR TITLE
refactor(run): use `HikesViewModel`

### PR DESCRIPTION
**Pull request by MaelImhof**
# Summary

As a complement to #241, I refactored the Run Hike screen (which was not ready to be modified when I started #241) for the entire app to use only `HikesViewModel`.

# Details

The new run hike screen is very *very* similar to the details screen. I introduced a composable that displays a loading screen and then provides the details of the input hike as output. This is shared logic between Run Hike and Hike Details, because they both need to load the hike's complete data before displaying it.

I refactored the Run Hike screen similar to how I did with the Hike Details screen, and I adapted the tests in a similar way too, because both screens are very similar.

### CODE REVIEW

**Review by mchac15 — CHANGES_REQUESTED** [review]
Really good PR and very well written code as always. 
I just found a very minor bug which should be solved by one line and since this is very similar code to what was done in the refactoring of the `HikeDetailsScreen` the code was already somewhat reviewed.

File: app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt:82]: At least it is necessary for the compiler. I believe the `LaunchedEffect` is a separate thing as the main body of the function, so it might get executed afterwards. Hence, both checks are needed.

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:80]: What this setup function does is :

- Load a list of saved hikes in the view model, as if the user went to the saved hikes screen
- Load (or not) its OSM data (waypoints and bounds mostly)
- Then load (or not) its elevation
- Then compute its details

The point being to have a completely loaded hike selected before the screen is loaded, so that information is displayed as expected.

However, to increase code coverage, I also checked the loading process. If the hike is not fully ready when the run hike is entered (which in practice is never the case, but code coverage...), the last details are loaded.

The first that is checked and loaded are the OSM details. The only way those are not loaded yet is for the hike to be only loaded from the saved hikes, and not from an OSM API call. Hence the choice of only accepting saved hikes. It's a rough simplification, but it shouldn't cause any issues to test only with saved hikes.

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:169]: Yup nice catch

File: app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt:105]: I'm not sure to understand what the observed bug is, could you maybe clarify?

`wantToNavigateBack`'s purpose is precisely to avoid spamming being a problem. With this, `navigationActions.goBack()` is called only once.

File: app/src/main/java/ch/hikemate/app/ui/components/WithDetailedHike.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/components/WithDetailedHike.kt:11]: Done 👌🏻

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:75]: I know I'm debating over a name, but I'm not sure the name you suggest describes what the parameter does. The parameter indicates whether the function should load the data into the view model (call `hikesViewModel.retrieveLoadedHikesOsmData()` and so on) or leave that to the test itself. You could say the loading is "forced", but that's not the point of the parameter, and it seems a bit misleading to me.

File: app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt:82]: I don't see why though since the `LaunchedEffect` will force the user to navigate back and thus it's technically impossible that this happens. 
I also don't see why this is true:
> At least it is necessary for the compiler.

File: app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt:105]: Sorry my explanation wasn't clear enough. Basically you can reproduce it doing the following:
Going very quickly between the `HikeDetailsScreen` and the `RunHikeScreen` by clicking on the _Run this hike_ button and _Stop the run_ **over and over** you'll see that if you do it quickly enough that the _Run this hike_ button won't work anymore after two-three times.
To solve that you need to force the value of the `wantToNavigateBack` and `wantToRunHike` to false after doing the action.
This is not the same problem as you stated but I should've explained it better.

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt
Comment (mchac15) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:80]: Oh I see and it makes sense, thanks for the explanation.

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt
Comment (mchac15) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:75]: You can leave it like that then. It isn't a big deal.

**Review by mchac15 — CHANGES_REQUESTED** [review]
Still a couple of discussions open and I hope the bug is clarified to you. If not I can personally send you a video reproducing it.
Thanks for the time taken!

File: app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt:82]: `LaunchedEffect(selectedHike) { ... }` only launches something, from what I understand we do not have any guarantee that the launched effect is executed first. It will be launched, not necessarily executed. This condition allows us to be certain 100% we do not have a null hike error.

File: app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt]: Ohhh I see. I was able to reproduce, I don't know why but I mistook this bug with another one, the one of the back button being spammed. Anyway, this should now be fixed !

File: app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt:82]: Plus, the `LaunchedEffect` does not stop the execution of the composable if the selected hike is null, it only calls `goBack`. That might take time to execute or something, at least the compiler is not happy with it, and I don't want to do an unsafe `!!`, so I check first.

Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/RunHikeScreen.kt:82]: Ok then thanks for the explanation. That makes sense.

**Review by mchac15 — APPROVED** [review]
Thanks for the changes!

**Review by mchac15 — COMMENTED** [review]
Once the CI passes and you merge to this branch you can merge to main

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt:1]: I just saw that you forgot to set to false the variable that launches the navigation to the RunHikeScreen which fixes the bug. Please add it.

**Review by mchac15 — CHANGES_REQUESTED** [review]
One little change more and we'll be done

**Review by mchac15 — APPROVED** [review]
LGTM now!
Once the CI passes you can merge

**Comment by MaelImhof** [comment]
I'm merging even though one of the CI does not pass. The CI don't pass because of the End2End 2 test. This test requires the elevation API to work, and the one we used until now has been kind of broken lately. @Wylwi is working on a new one in #247 but until I review it, the tests won't pass.

**Pull request merged**
